doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+## The Proposed Architecture
+
+### The Proposed Golang Package Structure
+
+Based on the foregoing analysis, the directory structure SQLFlow core code under `sqlflow/pkg` should be:
+
+| Package Name | Modification                                                 |
+| ------------ | ------------------------------------------------------------ |
+| `server`     | Calls `executor` instead of `workflow`.                      |
+| `proto`      | Add a new file: `ir.proto`.                                  |
+| `step`       | Remove `step/feature`. Move `tablewriter` to `step/tablewriter`.|
+| `parser`     | Stay the same.                                               |
+| `ir`         | Remove `ir/feature_column.go`.                               |
+| `executor`   | Rename`sql` to `executor`. Remove `sql/codegen` and `sql/codegen/attribute`. |
+| `database`   | Provide a python wrapper.                                    |
+| `pipe`       | Stay the same.                                               |
+| `log`        | Stay the same.                                               |
+
+### The Proposed Python Package Structure
+
+Add a package named `sqlflow` in [sqlflow/python](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/python) to contain the python code imgrated from golang. The directory structure of SQLFlow core code under `sqlflow/python/sqlflow` should be:
+
+| Package/Module Name | Description                                                  |
+| ------------------- | ------------------------------------------------------------ |
+| `platform`          | Modules under this package must implement the same set of functions(`run`, `train`, `evaluate`, `explain`, `predict` and `solve`) with the same signatures (e.g. `train(stmt: Statement)`) |
+| `columns`           | Modules under this package defines functions that can be called by users in a `COLUMN` clause. |
+| `features.py`       | Translate the `COLUMN` clause into calls to functions in the `column` package. |
+| `execute.py`        | Execute a SQLFlow program or a SQLFlow step.                 |
+| `client.py`         | A python client of sqlflow_server.                           |
+| `_database.so`      | Wrapper of the golang package `pkg/database`.                |
+| `database.py`       | Wrapper to `_database.so` for ease of use.                   |
+| `ir_pb2.py`         | Generated from `pkg/proto/ir.proto`.                         |
+| `contracts.py`      | See below.                                                   |
+| `diagnostics.py`    | See below.                                                   |
+
+#### Python Version And The Style
+
+The `sqlflow` package will only run in the *Docker container* and never be submitted to the PAI platform, so we propose to develop in python3 in the migration.
```

Comment:
Le's take tensorflow for example, the transform function for `column` clause will call the keras preprocessing layers or feature column api. It's a part of the complete model definition and will run on the PAI platform.

Reply:
You're right. I'll fix this later.